### PR TITLE
Feature/const

### PR DESCRIPTION
**Pull request by lelemmen**
This PR implements the const-correct feature we've been dreaming about. See #42.

### CODE REVIEW

**Comment by lelemmen** [comment]
It also implements #161, #169, #171 and #147.

**Comment by tmhuysen** [comment]
Some of the optimization tests are failing at random, some of the times.

Issues appear to be minor accuracy changes.
eg. 0.999999999996 for the norm of an eigenvector. This fails the 10e-12 threshold (would succeed for 10e-11)

I've been testing a local old numopt code with the same parameters:
>gqcp compiled "-g -march=native, EIGEN3 3.3.5"
>numopt compiled "-g -march=native, EIGEN3 3.3.5"

The test that is ran : "liu_50_collapse"

The gqcp code fails approximately 1/2 of the times.
The numopt code does not fail.

I've been gradually changing numopt code to that of current live GQCP version. No fails so far.

File: include/optimization/BaseEigenproblemSolver.hpp [hunk 1]
```diff
@@ -59,50 +55,30 @@ class BaseEigenproblemSolver {
...
     // GETTERS
-    /**
-     *  @return the diagonal of the matrix representation of the matrix whose eigenproblem is being solved
-     */
-    virtual Eigen::VectorXd get_diagonal() = 0;
     bool is_solved() const { return this->_is_solved; }
 
-
-    // GETTERS - EIGENPAIR
     std::vector<Eigenpair> get_eigenpairs() const;
```
Comment (tmhuysen) [inline include/optimization/BaseEigenproblemSolver.hpp]: This line should return a const reference not a copy, this causes the sudden drop in accuracy.

Comment (lelemmen) [reply include/optimization/BaseEigenproblemSolver.hpp]: Sorry, my bad. Thanks for catching this!

**Pull request merged**